File: union_find/restricted_requests/restricted_requests.py

class RestrictedFriendRequests:  # LeetCode Q.2076.
    def __init__(self):
        self.parents: list[int] = []
        self.forbidden: dict[int, set[int]] = {}

    def process_requests(
            self, n: int, restrictions: list[list[int]], requests: list[list[int]]
    ) -> list[bool]:
        self.parents.clear()  # Initialize each person's parent as self.
        self.parents = [person for person in range(n)]

        self.forbidden.clear()  # Each person's restricted people.
        self.forbidden.update(
            dict(
                zip([person for person in range(n)], [set() for _ in range(n)])
            )
        )

        for person_1, person_2 in restrictions:
            self.forbidden[person_1].add(person_2)
            self.forbidden[person_2].add(person_1)

        requests_results = []
        for person_1, person_2 in requests:
            parent_1 = self._find_parent(person_1)
            parent_2 = self._find_parent(person_2)
            request_success = True  # Default to True.

            for forbidden_person in self.forbidden[parent_1]:  # Test parent 1.
                if self._find_parent(forbidden_person) == parent_2:
                    request_success = False
                    break
            if request_success:  # Need to test parent 2.
                for forbidden_person in self.forbidden[parent_2]:
                    if self._find_parent(forbidden_person) == parent_1:
                        request_success = False
                        break

            requests_results.append(request_success)
            if request_success:
                self._union_people(parent_1, parent_2)

        return requests_results

    def _find_parent(self, person: int) -> int:
        parent = self.parents[person]
        return person if parent == person else self._find_parent(parent)

    def _union_people(self, person_1: int, person_2: int) -> None:
        parent_1 = self._find_parent(min(person_1, person_2))
        parent_2 = self._find_parent(max(person_1, person_2))
        if parent_1 != parent_2:
            self.parents[parent_2] = parent_1

            # Must merge parent 2's restricted people into parent 1's.
            self.forbidden[parent_1].update(self.forbidden[parent_2])
            del self.forbidden[parent_2]  # Clear unused memory.

File: union_find/restricted_requests/restricted_requests.cpp
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

class RestrictedFriendRequests
{ // LeetCode Q.2076.
private:
    // Each person's restricted people.
    unordered_map<int, unordered_set<int>> forbidden;
    vector<int> parents;

    int findParent(int person)
    {
        if (parents[person] == person)
            return person;

        return findParent(parents[person]);
    }

    void unionPeople(int personOne, int personTwo)
    {
        int parentOne = findParent(min(personOne, personTwo));
        int parentTwo = findParent(max(personOne, personTwo));

        if (parentOne != parentTwo)
        {
            parents[parentTwo] = parentOne;

            // Must merge parent 2's restricted people into parent 1's.
            forbidden[parentOne].insert(
                forbidden[parentTwo].begin(), forbidden[parentTwo].end());

            forbidden.erase(parentTwo); // Clear unused memory.
        }
    }

public:
    vector<bool> processRequests(int n, vector<vector<int>> &restrictions, vector<vector<int>> &requests)
    {
        for (int person = 0; person < n; person++)
        {
            parents.push_back(person); // Initialize each person's parent as self.
            forbidden[person] = {};    // Initialize empty set.
        }

        for (auto restriction : restrictions)
        {
            forbidden[restriction[0]].insert(restriction[1]);
            forbidden[restriction[1]].insert(restriction[0]);
        }

        vector<bool> requestsResults;

        for (auto request : requests)
        {
            int parentOne = findParent(request[0]);
            int parentTwo = findParent(request[1]);
            bool requestSuccess = true; // Default to True.

            for (auto forbidden_person : forbidden[parentOne]) // Test parent 1.
            {
                if (findParent(forbidden_person) == parentTwo)
                {
                    requestSuccess = false;
                    break;
                }
            }

            if (requestSuccess) // Test parent 2.
            {
                for (auto forbidden_person : forbidden[parentTwo])
                {
                    if (findParent(forbidden_person) == parentOne)
                    {
                        requestSuccess = false;
                        break;
                    }
                }
            }

            requestsResults.push_back(requestSuccess);
            if (requestSuccess)
                unionPeople(parentOne, parentTwo);
        }

        return requestsResults;
    }
};